subprojects/base-services/src/main/java/org/gradle/internal/service/DefaultServiceRegistry.java

```diff
@@ -414,27 +402,31 @@ private static Type extractServiceType(Type mayBeFactoryType) {
          * Collects all services of the given type.
          */
         void getAll(LookupContext context, Class<?> serviceType, List<ServiceProvider> result);
-
-        boolean hasService(Class<?> type);
     }
 
     private class OwnServices implements Provider {
-        private List<Provider> providers;
-        private Set<Class<?>> serviceTypes;
+        private final Map<Class<?>, Collection<Provider>> providersByType = new HashMap<Class<?>, Collection<Provider>>();
+        private final CompositeStoppable stoppable = CompositeStoppable.stoppable();
 
         @Override
         public ServiceProvider getFactory(LookupContext context, Class<?> type) {
-            if (!hasService(Factory.class)) {
+            Collection<Provider> providers = getProvidersByType(Factory.class);
+            if (providers.isEmpty()) {
                 return null;
             }
-            List<ServiceProvider> candidates = new ArrayList<ServiceProvider>();
+            if (providers.size() == 1) {
+                return providers.iterator().next().getFactory(context, type);
```

Comment:
:o: We can avoid an iterator here by strongly typing providers to `ArrayList`. This is always the case. Then you can get the single provider in constant time `.get(0)`

Reply:
Right, this is leftover from when I used a `Set`, but that's no covered by the type hierarchy walking.

Reply:
Fixed